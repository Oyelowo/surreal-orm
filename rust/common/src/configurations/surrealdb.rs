use super::utils::get_env_vars_by_prefix;
use serde::Deserialize;
use serde_aux::prelude::deserialize_number_from_string;
use surrealdb::Datastore;

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub struct SurrealdbConfigs {
    pub namespace: String,
    pub database: String,
    pub username: String,
    pub password: String,
    pub host: String,

    #[serde(deserialize_with = "deserialize_number_from_string")]
    pub port: u16,

    #[serde(default = "default_require_ssl")]
    pub require_ssl: Option<bool>,
}

fn default_require_ssl() -> Option<bool> {
    Some(false)
}

impl Default for SurrealdbConfigs {
    fn default() -> Self {
        get_env_vars_by_prefix("SURREALDB_")
    }
}

impl SurrealdbConfigs {
    pub fn get_database(self) -> anyhow::Result<Datastore> {
        // let ds = Datastore::new("tikv://127.0.0.1:8000").await?;
        todo!()
    }
}
