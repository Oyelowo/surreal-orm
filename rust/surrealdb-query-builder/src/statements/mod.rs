use std::fmt::{Display, Formatter};

pub(crate) mod chain;
pub(crate) mod create;
pub(crate) mod define_database;
pub(crate) mod define_event;
pub(crate) mod define_field;
pub(crate) mod define_index;
pub(crate) mod define_login;
pub(crate) mod define_namespace;
pub(crate) mod define_scope;
pub(crate) mod define_table;
pub(crate) mod define_token;
pub(crate) mod delete;
pub(crate) mod for_;
pub(crate) mod ifelse;
pub(crate) mod info;
pub(crate) mod insert;
pub(crate) mod let_;
pub(crate) mod relate;
pub(crate) mod remove_database;
pub(crate) mod remove_event;
pub(crate) mod remove_field;
pub(crate) mod remove_index;
pub(crate) mod remove_login;
pub(crate) mod remove_namespace;
pub(crate) mod remove_scope;
pub(crate) mod remove_table;
pub(crate) mod remove_token;
pub(crate) mod select;
pub(crate) mod sleep;
pub(crate) mod transaction;
pub(crate) mod update;
pub(crate) mod use_;

pub mod statements {
    pub use super::chain::{chain, QueryChain};
    pub use super::create::{create, CreateStatement};
    pub use super::define_database::{define_database, DefineDatabaseStatement};
    pub use super::define_event::{define_event, DefineEventStatement};
    pub use super::define_field::{define_field, DefineFieldStatement, FieldType};
    pub use super::define_index::{define_index, DefineIndexStatement};
    pub use super::define_login::{define_login, DefineLoginStatement};
    pub use super::define_namespace::{define_namespace, DefineNamespaceStatement};
    pub use super::define_scope::{define_scope, DefineScopeStatement};
    pub use super::define_table::{define_table, DefineTableStatement};
    pub use super::define_token::{define_token, DefineTokenStatement};
    pub use super::delete::{delete, DeleteStatement};
    pub use super::for_::{for_, For, ForCrudType};
    pub use super::ifelse::{if_, IfStatement};
    pub use super::info::{info_for, InfoStatement};
    pub use super::insert::{insert, InsertStatement, Insertables};
    pub use super::let_::{let_, LetStatement};
    pub use super::relate::{relate, RelateStatement};
    pub use super::remove_database::{remove_database, RemoveDatabaseStatement};
    pub use super::remove_event::{remove_event, RemoveEventStatement};
    pub use super::remove_field::{remove_field, RemoveFieldStatement};
    pub use super::remove_index::{remove_index, RemoveIndexStatement};
    pub use super::remove_login::{remove_login, RemoveLoginStatement};
    pub use super::remove_namespace::{remove_namespace, RemoveNamespaceStatement};
    pub use super::remove_scope::{remove_scope, RemoveScopeStatement};
    pub use super::remove_table::{remove_table, RemoveTableStatement};
    pub use super::remove_token::{remove_token, RemoveTokenStatement};
    pub use super::select::{select, SelectStatement};
    pub use super::sleep::{sleep, SleepStatement};
    pub use super::transaction::{begin_transaction, BeginTransactionStatement};
    pub use super::update::{update, UpdateStatement};
    pub use super::use_::{use_, UseStatement};
}

pub(crate) enum NamespaceOrDatabase {
    Namespace,
    Database,
}

impl Display for NamespaceOrDatabase {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let stringified = match self {
            NamespaceOrDatabase::Namespace => "NAMESPACE",
            NamespaceOrDatabase::Database => "DATABASE",
        };
        write!(f, "{}", stringified)
    }
}
