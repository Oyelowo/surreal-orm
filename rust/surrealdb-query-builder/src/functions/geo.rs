/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowooyedayo@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

// Geo functions
// These functions can be used when working with and analysing geospatial data.
//
// Function	Description
// geo::area()	Calculates the area of a geometry
// geo::bearing()	Calculates the bearing between two geolocation points
// geo::centroid()	Calculates the centroid of a geometry
// geo::distance()	Calculates the distance between two geolocation points
// geo::hash::decode()	Decodes a geohash into a geometry point
// geo::hash::encode()	Encodes a geometry point into a geohash
//
//

use geo::{point, polygon};
use surrealdb::sql;

use crate::{
    field::GeometryOrField,
    sql::{Binding, Buildable, Empty, ToRawStatement},
    Field,
};

use super::array::Function;

// enum GeometryOrField {
//     Field(Field),
//     Geometry(sql::Geometry),
// }
//
// impl From<GeometryOrField> for sql::Value {
//     fn from(value: GeometryOrField) -> Self {
//         match value {
//             GeometryOrField::Field(f) => f.into(),
//             GeometryOrField::Geometry(geo) => geo.into(),
//         }
//     }
// }
//
// impl From<Field> for GeometryOrField {
//     fn from(value: Field) -> Self {
//         Self::Field(value)
//     }
// }
//
// impl<T> From<T> for GeometryOrField
// where
//     T: Into<sql::Geometry>,
// {
//     fn from(value: T) -> Self {
//         let value: sql::Geometry = value.into();
//         Self::Geometry(value)
//     }
// }
pub(crate) fn create_geo_with_single_arg(
    geometry: impl Into<GeometryOrField>,
    fn_suffix: &str,
) -> Function {
    let binding = Binding::new(geometry.into());
    let string = binding.get_param_dollarised();

    Function {
        query_string: format!("geo::{fn_suffix}({})", string),
        bindings: vec![binding],
    }
}

fn create_geo_fn_with_two_args(
    point1: impl Into<GeometryOrField>,
    point2: impl Into<GeometryOrField>,
    fn_suffix: &str,
) -> Function {
    let binding1 = Binding::new(point1.into());
    let string1 = binding1.get_param_dollarised();

    let binding2 = Binding::new(point2.into());
    let string2 = binding2.get_param_dollarised();

    Function {
        query_string: format!("geo::{fn_suffix}({}, {})", string1, string2),
        bindings: vec![binding1, binding2],
    }
}

pub fn area_fn(geometry: impl Into<GeometryOrField>) -> Function {
    create_geo_with_single_arg(geometry, "area")
}

#[macro_export]
macro_rules! geo_area {
    ( $geometry:expr ) => {
        crate::functions::geo::area_fn($geometry)
    };
}

pub use geo_area as area;

pub fn bearing_fn(
    point1: impl Into<GeometryOrField>,
    point2: impl Into<GeometryOrField>,
) -> Function {
    create_geo_fn_with_two_args(point1, point2, "bearing")
}

#[macro_export]
macro_rules! geo_bearing {
    ( $point1:expr,  $point2:expr ) => {
        crate::functions::geo::area_fn($point1, $point2)
    };
}

pub use geo_bearing as bearing;

pub fn centroid(geometry: impl Into<GeometryOrField>) -> Function {
    create_geo_with_single_arg(geometry, "centroid")
}

#[macro_export]
macro_rules! geo_centroid {
    ( $geometry:expr ) => {
        crate::functions::geo::centroid_fn($geometry)
    };
}

pub use geo_centroid as centroid;

pub fn distance_fn(
    point1: impl Into<GeometryOrField>,
    point2: impl Into<GeometryOrField>,
) -> Function {
    create_geo_fn_with_two_args(point1, point2, "distance")
}

#[macro_export]
macro_rules! geo_distance {
    ( $point1:expr,  $point2:expr ) => {
        crate::functions::geo::distance_fn($point1, $point2)
    };
}
pub use geo_distance as distance;

pub enum NumberOrEmpty {
    Empty,
    Number(surrealdb::sql::Value),
    // Field(Field),
}

// impl From<NumberOrEmpty> for sql::Value {
//     fn from(value: NumberOrEmpty) -> Self {
//         match value {
//             NumberOrEmpty::Empty => sql::Idiom::from("".to_string()).into(),
//             NumberOrEmpty::Number(n) => n.into(),
//             // NumberOrEmpty::Field(f) => f.into(),
//         }
//     }
// }
impl<T> From<T> for NumberOrEmpty
where
    T: Into<sql::Number>,
{
    fn from(value: T) -> Self {
        let value: sql::Number = value.into();
        Self::Number(value.into())
    }
}

impl From<Field> for NumberOrEmpty {
    fn from(value: Field) -> Self {
        Self::Number(value.into())
    }
}

impl From<Empty> for NumberOrEmpty {
    fn from(value: Empty) -> Self {
        Self::Empty
    }
}

pub mod hash {

    type Accuracy = super::NumberOrEmpty;

    use surrealdb::sql;

    use super::create_geo_with_single_arg;
    use crate::{
        field::GeometryOrField,
        functions::array::Function,
        sql::{Binding, Empty},
        Field,
    };

    pub struct GeoHash(sql::Value);

    impl From<&str> for GeoHash {
        fn from(value: &str) -> Self {
            Self(value.into())
        }
    }
    impl From<String> for GeoHash {
        fn from(value: String) -> Self {
            Self(value.into())
        }
    }

    impl From<Field> for GeoHash {
        fn from(value: Field) -> Self {
            Self(value.into())
        }
    }

    impl From<GeoHash> for sql::Value {
        fn from(value: GeoHash) -> Self {
            value.0
        }
    }

    // impl GeoHash {
    //     fn new(hash: String) -> Self {
    //         Self(hash)
    //     }
    // }
    //
    // enum GeoHashDecodeArg {
    //     Field(Field),
    //     GeoHash(GeoHash),
    // }

    pub fn decode(geometry: impl Into<GeoHash>) -> Function {
        let binding = Binding::new(geometry.into());
        let string = binding.get_param_dollarised();

        Function {
            query_string: format!("geo::hash::decode({})", string),
            bindings: vec![binding],
        }
    }
    #[macro_export]
    macro_rules! geo_hash_decode {
        ( $geometry:expr ) => {
            crate::functions::geo::decode_fn($geometry)
        };
    }
    pub use geo_hash_decode as decode;

    pub fn encode(geometry: impl Into<GeometryOrField>, accuracy: impl Into<Accuracy>) -> Function {
        let binding = Binding::new(geometry.into());
        let accuracy: Accuracy = accuracy.into();
        let geometry_param = binding.get_param_dollarised();

        let mut bindings = vec![binding];

        let str = match accuracy {
            Accuracy::Empty => format!("geo::hash::encode({geometry_param})",),
            Accuracy::Number(num) => {
                let binding = Binding::new(num);
                let accuracy_param = binding.get_param_dollarised();
                bindings.push(binding);

                format!("geo::hash::encode({geometry_param}, {accuracy_param})",)
            } // Accuracy::Field(field) => {
              //     let binding = Binding::new(field);
              //     let accuracy_param = binding.get_param_dollarised();
              //     bindings.push(binding);
              //
              //     format!("geo::hash::encode({geometry_param}, {accuracy_param})",)
              // }
        };
        Function {
            query_string: str,
            bindings,
        }
    }

    #[macro_export]
    macro_rules! geo_hash_encode {
        ( $geometry:expr, $accuracy:expr ) => {
            crate::functions::geo::encode_fn($geometry, $accuracy)
        };
        ( $geometry:expr ) => {
            crate::functions::geo::encode_fn($geometry)
        };
    }
    pub use geo_hash_encode as encode;
}
#[test]
fn test_area_with_field() {
    let city = Field::new("city");
    let result = area_fn(city);

    assert_eq!(result.fine_tune_params(), "geo::area($_param_00000001)");
    assert_eq!(result.to_raw().to_string(), "geo::area(city)");
}

#[test]
fn test_area_with_raw_polygon() {
    let poly = polygon!(
            exterior: [
                (x: -111., y: 45.),
                (x: -111., y: 41.),
                (x: -104., y: 41.),
                (x: -104., y: 45.),
            ],
            interiors: [
                [
                    (x: -110., y: 44.),
                    (x: -110., y: 42.),
                    (x: -105., y: 42.),
                    (x: -105., y: 44.),
                ],
            ],
        );
    let result = area_fn(poly);
    assert_eq!(
        result.fine_tune_params(),
        "geo::area($_param_00000001)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::area({ type: 'Polygon', coordinates: [[[-111, 45], [-111, 41], [-104, 41], [-104, 45], [-111, 45]], [[[-110, 44], [-110, 42], [-105, 42], [-105, 44], [-110, 44]]]] })"
    );
}

#[test]
fn test_area_with_raw_polygon() {
    let poly = polygon!(
            exterior: [
                (x: -111., y: 45.),
                (x: -111., y: 41.),
                (x: -104., y: 41.),
                (x: -104., y: 45.),
            ],
            interiors: [
                [
                    (x: -110., y: 44.),
                    (x: -110., y: 42.),
                    (x: -105., y: 42.),
                    (x: -105., y: 44.),
                ],
            ],
        );
    let result = area_fn(poly);
    assert_eq!(
        result.fine_tune_params(),
        "geo::area($_param_00000001)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::area({ type: 'Polygon', coordinates: [[[-111, 45], [-111, 41], [-104, 41], [-104, 45], [-111, 45]], [[[-110, 44], [-110, 42], [-105, 42], [-105, 44], [-110, 44]]]] })"
    );
}
#[test]
fn test_bearing_with_raw_points() {
    let point1 = point! {
        x: 40.02f64,
        y: 116.34,
    };

    let point2 = point! {
        x: 80.02f64,
        y: 103.19,
    };
    let result = bearing_fn(point1, point2);
    assert_eq!(
        result.fine_tune_params(),
        "geo::bearing($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::bearing((40.02, 116.34), (80.02, 103.19))"
    );
}

#[test]
fn test_bearing_with_raw_point_with_field() {
    let hometown = Field::new("hometown");

    let point2 = point! {
        x: 80.02f64,
        y: 103.19,
    };
    let result = bearing_fn(hometown, point2);
    assert_eq!(
        result.fine_tune_params(),
        "geo::bearing($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::bearing(hometown, (80.02, 103.19))"
    );
}

#[test]
fn test_centroid_with_field() {
    let city = Field::new("city");
    let result = centroid(city);

    assert_eq!(result.fine_tune_params(), "geo::centroid($_param_00000001)");
    assert_eq!(result.to_raw().to_string(), "geo::centroid(city)");
}

#[test]
fn test_centroid_with_raw_polygon() {
    let poly = polygon!(
            exterior: [
                (x: -111., y: 45.),
                (x: -111., y: 41.),
                (x: -104., y: 41.),
                (x: -104., y: 45.),
            ],
            interiors: [
                [
                    (x: -110., y: 44.),
                    (x: -110., y: 42.),
                    (x: -105., y: 42.),
                    (x: -105., y: 44.),
                ],
            ],
        );
    let result = centroid(poly);
    assert_eq!(
        result.fine_tune_params(),
        "geo::centroid($_param_00000001)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::centroid({ type: 'Polygon', coordinates: [[[-111, 45], [-111, 41], [-104, 41], [-104, 45], [-111, 45]], [[[-110, 44], [-110, 42], [-105, 42], [-105, 44], [-110, 44]]]] })"
    );
}

#[test]
fn test_distance_with_raw_points() {
    let point1 = point! {
        x: 40.02f64,
        y: 116.34,
    };

    let point2 = point! {
        x: 80.02f64,
        y: 103.19,
    };
    let result = distance_fn(point1, point2);
    assert_eq!(
        result.fine_tune_params(),
        "geo::distance($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::distance((40.02, 116.34), (80.02, 103.19))"
    );
}

#[test]
fn test_distance_with_raw_point_with_field() {
    let hometown = Field::new("hometown");

    let point2 = point! {
        x: 80.02f64,
        y: 103.19,
    };
    let result = distance_fn(hometown, point2);
    assert_eq!(
        result.fine_tune_params(),
        "geo::distance($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::distance(hometown, (80.02, 103.19))"
    );
}

#[test]
fn test_distance_with_only_fields() {
    let hometown = Field::new("hometown");
    let yukon = Field::new("yukon");

    let result = distance_fn(hometown, yukon);
    assert_eq!(
        result.fine_tune_params(),
        "geo::distance($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::distance(hometown, yukon)"
    );
}

#[test]
fn test_hash_decode_with_field() {
    let city = Field::new("city");
    let result = hash::decode(city);

    assert_eq!(
        result.fine_tune_params(),
        "geo::hash::decode($_param_00000001)"
    );
    assert_eq!(result.to_raw().to_string(), "geo::hash::decode(city)");
}

#[test]
fn test_hash_decode_with_string() {
    let result = hash::decode("mpuxk4s24f51");
    assert_eq!(
        result.fine_tune_params(),
        "geo::hash::decode($_param_00000001)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::hash::decode('mpuxk4s24f51')"
    );
}

#[test]
fn test_hash_encode_with_field_and_empty_accuracy() {
    let city = Field::new("city");
    let result = hash::encode(city, Empty);

    assert_eq!(
        result.fine_tune_params(),
        "geo::hash::encode($_param_00000001)"
    );
    assert_eq!(result.to_raw().to_string(), "geo::hash::encode(city)");
}

#[test]
fn test_hash_encode_with_field_and_field_accuracy() {
    let city = Field::new("city");
    let accuracy = Field::new("accuracy");
    let result = hash::encode(city, accuracy);

    assert_eq!(
        result.fine_tune_params(),
        "geo::hash::encode($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::hash::encode(city, accuracy)"
    );
}

#[test]
fn test_hash_encode_with_field_and_number_accuracy() {
    let city = Field::new("city");
    let result = hash::encode(city, 5);

    assert_eq!(
        result.fine_tune_params(),
        "geo::hash::encode($_param_00000001, $_param_00000002)"
    );
    assert_eq!(result.to_raw().to_string(), "geo::hash::encode(city, 5)");
}

#[test]
fn test_hash_encode_with_point() {
    let point = point! {
        x: 40.02f64,
        y: 116.34,
    };

    let result = hash::encode(point, 5);
    assert_eq!(
        result.fine_tune_params(),
        "geo::hash::encode($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "geo::hash::encode((40.02, 116.34), 5)"
    );
}
