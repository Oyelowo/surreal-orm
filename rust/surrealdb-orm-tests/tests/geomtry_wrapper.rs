/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowooyedayo@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

use geo::LineString;
use geo::MultiLineString;
use geo::MultiPoint;
use geo::MultiPolygon;
use geo::Point;
use geo::Polygon;

use geo::point;
use geo::polygon;
use serde::Deserialize;
use serde::Serialize;
use surrealdb::engine::local::Mem;
use surrealdb::opt::RecordId;
use surrealdb::sql;
use surrealdb::sql::Datetime;
// use surrealdb::sql::Geometry;
use surrealdb::sql::Uuid;
use surrealdb::Surreal;
use surrealdb_orm::Buildable;
use surrealdb_orm::RunnableSelect;
use surrealdb_orm::ToRaw;
// use surrealdb_derive::SurrealdbNode;
use std::time::Duration;

use geo::Coord;
use surrealdb::sql::statements::CommitStatement;
use surrealdb_orm::{
    statements::{insert, select},
    All, Geometry, Operatable, Parametric, Runnable, RunnableDefault, RunnableStandard, SurrealId,
    SurrealdbNode,
};

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Person {
    name: String,
}
#[derive(SurrealdbNode, Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
#[surrealdb(table_name = "company")]
struct Company {
    #[serde(skip_serializing_if = "Option::is_none")]
    // #[builder(default, setter(strip_option))]
    id: Option<SurrealId>,
    nam: Uuid,
    name: String,
    founded: Datetime,
    founders: Vec<Person>,
    tags: Vec<String>,
    home: sql::Geometry,
}

#[derive(SurrealdbNode, Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
#[surrealdb(table_name = "gen_z_company")]
struct GenZCompany {
    #[serde(skip_serializing_if = "Option::is_none")]
    // #[builder(default, setter(strip_option))]
    id: Option<SurrealId>,
    nam: Uuid,
    name: String,
    founded: Datetime,
    founders: Vec<Person>,
    tags: Vec<String>,
    home: sql::Geometry,
}

#[derive(SurrealdbNode, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surrealdb(table_name = "book")]
pub struct Book {
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<SurrealId>,
    title: String,
    content: String,
}

fn create_test_company(geom: impl Into<sql::Geometry>) -> Company {
    let company = Company {
        id: Some(RecordId::from(("company", "lowo")).into()),
        nam: Uuid::try_from("285cfebe-a7f2-4100-aeb3-7f73998fff02").unwrap(),
        name: "Mana Inc.".to_string(),
        founded: "1967-05-03".into(),
        founders: vec![
            Person {
                name: "John Doe".to_string(),
            },
            Person {
                name: "Jane Doe".to_string(),
            },
        ],
        tags: vec!["foo".to_string(), "bar".to_string()],
        home: geom.into(),
    };
    company
}

async fn create_geom_test(geom: impl Into<sql::Geometry>) -> surrealdb::Result<String> {
    let company = create_test_company(geom);
    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test").use_db("test").await?;

    // let results = insert::<Company>(company);
    let results = insert(company).return_one(db).await.unwrap().unwrap();

    Ok(serde_json::to_string(&results).unwrap())
}

#[tokio::test]
async fn point() -> surrealdb::Result<()> {
    let point = point! {
        x: 40.02f64,
        y: 116.34,
    };

    // let company = create_geom_test(point).await?;
    // insta::assert_snapshot!(company);
    println!("sql::Geom {:?}", sql::Geometry::Point(point.into()));
    println!("Geom basic{:?}", point);
    println!("Geom own{:?}", Geometry(point.into()));
    // insta::assert_debug_snapshot!(sql::Geometry::Point(point.into()));
    Ok(())
}

#[tokio::test]
async fn linestring() -> surrealdb::Result<()> {
    let ls = LineString(vec![
        Coord {
            x: -122.33583,
            y: 47.60621,
        },
        Coord {
            x: -122.33583,
            y: 47.60622,
        },
        Coord {
            x: -122.33584,
            y: 47.60622,
        },
        Coord {
            x: -122.33584,
            y: 47.60621,
        },
        Coord {
            x: -122.33583,
            y: 47.60621,
        },
    ]);

    let company = create_geom_test(ls).await?;
    insta::assert_snapshot!(company);
    Ok(())
}

#[tokio::test]
async fn polygon() -> surrealdb::Result<()> {
    let polygon = polygon![
            (x: -111., y: 45.),
            (x: -111., y: 41.),
            (x: -104., y: 41.),
            (x: -104., y: 45.),
        // (x: 0.0, y: 0.0),
        // (x: 4.0, y: 0.0),
        // (x: 4.0, y: 1.0),
        // (x: 1.0, y: 1.0),
        // (x: 1.0, y: 4.0),
        // (x: 0.0, y: 4.0),
        // (x: 0.0, y: 0.0),
    ];
    //
    let company = create_geom_test(polygon).await?;
    insta::assert_snapshot!(company);

    let poly = polygon!(
            exterior: [
                (x: -111., y: 45.),
                (x: -111., y: 41.),
                (x: -104., y: 41.),
                (x: -104., y: 45.),
            ],
            interiors: [
                [
                    (x: -110., y: 44.),
                    (x: -110., y: 42.),
                    (x: -105., y: 42.),
                    (x: -105., y: 44.),
                ],
            ],
        );

    let company_complex = create_geom_test(poly).await?;
    println!(
        "ZMZMZMZM {}",
        serde_json::to_string(&company_complex).unwrap()
    );
    insta::assert_snapshot!(company_complex);
    Ok(())
}

#[tokio::test]
async fn multipoint() -> surrealdb::Result<()> {
    let points = MultiPoint(vec![
        Point::new(0.0, 0.0),
        Point::new(1.0, 1.0),
        (2.0, 35.0).into(),
    ]);

    let company = create_geom_test(points).await?;
    insta::assert_snapshot!(company);
    Ok(())
}

#[tokio::test]
async fn multiline() -> surrealdb::Result<()> {
    let linestring1 = LineString(vec![
        Coord { x: 0.0, y: 0.0 },
        Coord { x: 1.0, y: 1.0 },
        Coord { x: 2.0, y: 2.0 },
    ]);
    let linestring2 = LineString(vec![
        Coord { x: 3.0, y: 3.0 },
        Coord { x: 4.0, y: 4.0 },
        Coord { x: 5.0, y: 5.0 },
    ]);
    let multiline_string = MultiLineString(vec![linestring1, linestring2]);

    let company = create_geom_test(multiline_string).await?;
    insta::assert_snapshot!(company);
    Ok(())
}

#[tokio::test]
async fn multipolygon() -> surrealdb::Result<()> {
    let polygon1 = Polygon::new(
        LineString(vec![
            Coord { x: 0.0, y: 0.0 },
            Coord { x: 1.0, y: 1.0 },
            Coord { x: 2.0, y: 2.0 },
            Coord { x: 0.0, y: 0.0 },
        ]),
        vec![],
    );
    let polygon2 = Polygon::new(
        LineString(vec![
            Coord { x: 3.0, y: 3.0 },
            Coord { x: 4.0, y: 4.0 },
            Coord { x: 5.0, y: 5.0 },
            Coord { x: 3.0, y: 3.0 },
        ]),
        vec![],
    );
    let poly3 = polygon!(
            exterior: [
                (x: -111., y: 45.),
                (x: -111., y: 41.),
                (x: -104., y: 41.),
                (x: -104., y: 45.),
            ],
            interiors: [
                [
                    (x: -110., y: 44.),
                    (x: -110., y: 42.43),
                    (x: -105., y: 42.),
                    (x: -105., y: 44.),
                ],
            ],
        );
    let multi_polygon = MultiPolygon(vec![polygon1, polygon2, poly3]);
    insta::assert_snapshot!(serde_json::to_string(&multi_polygon).unwrap());
    let company = create_geom_test(multi_polygon).await?;
    insta::assert_snapshot!(company);
    Ok(())
}

#[tokio::test]
async fn geom_collection() -> surrealdb::Result<()> {
    let point = Point(Coord { x: 0.0, y: 0.0 });
    let linestring = LineString(vec![Coord { x: 1.0, y: 1.0 }, Coord { x: 2.0, y: 2.0 }]);
    let geometry_collection = vec![sql::Geometry::Point(point), sql::Geometry::Line(linestring)];
    let geometry_collection = sql::Geometry::Collection(geometry_collection);
    let company = create_geom_test(geometry_collection).await?;
    insta::assert_snapshot!(company);
    Ok(())
}

#[tokio::test]
async fn insert_many() -> surrealdb::Result<()> {
    let companies = vec![
        Company {
            id: Some("company:1".try_into().unwrap()),
            name: "Acme Inc.".to_string(),
            founded: "1967-05-03".into(),
            founders: vec![
                Person {
                    name: "John Doe".to_string(),
                },
                Person {
                    name: "Jane Doe".to_string(),
                },
            ],
            tags: vec!["foo".to_string(), "bar".to_string()],
            nam: Uuid::try_from("725cfebe-a7f2-4100-aeb3-7f73998fff02").unwrap(),
            home: (45.3, 78.1).into(),
        },
        Company {
            id: Some("company:2".try_into().unwrap()),
            name: "Apple Inc.".to_string(),
            founded: "1967-05-03".into(),
            founders: vec![
                Person {
                    name: "John Doe".to_string(),
                },
                Person {
                    name: "Jane Doe".to_string(),
                },
            ],
            tags: vec!["foo".to_string(), "bar".to_string()],
            nam: Uuid::try_from("375cfebe-a7f2-4100-aeb3-7f73998fff02").unwrap(),
            home: (63.0, 21.0).into(),
        },
    ];

    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test").use_db("test").await?;

    let results = insert(companies).return_many(db).await.unwrap();

    insta::assert_debug_snapshot!(results);
    Ok(())
}

#[tokio::test]
async fn insert_from_select_query() -> surrealdb::Result<()> {
    let companies = vec![
        Company {
            id: Some("company:1".try_into().unwrap()),
            name: "Acme Inc.".to_string(),
            founded: "1967-05-03".into(),
            founders: vec![
                Person {
                    name: "John Doe".to_string(),
                },
                Person {
                    name: "Jane Doe".to_string(),
                },
            ],
            tags: vec!["foo".to_string(), "bar".to_string()],
            nam: Uuid::try_from("725cfebe-a7f2-4100-aeb3-7f73998fff02").unwrap(),
            home: (45.3, 78.1).into(),
        },
        Company {
            id: Some("company:2".try_into().unwrap()),
            name: "Apple Inc.".to_string(),
            founded: "1967-05-03".into(),
            founders: vec![
                Person {
                    name: "John Doe".to_string(),
                },
                Person {
                    name: "Jane Doe".to_string(),
                },
            ],
            tags: vec!["foo".to_string(), "bar".to_string()],
            nam: Uuid::try_from("375cfebe-a7f2-4100-aeb3-7f73998fff02").unwrap(),
            home: (63.0, 21.0).into(),
        },
    ];

    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test").use_db("test").await?;
    // Insert companies
    // let results = insert(companies).return_many(db.clone()).await.unwrap();
    let results = insert(companies).return_many(db.clone()).await.unwrap();
    // results.into_iter().collect();

    // let results = insert(companies)
    //     .return_many(db.clone())
    //     // .return_many_test::<Vec<_>>(db.clone())
    //     .await
    //     .unwrap();

    println!("QQQQQQINS {:?}", results);
    // db.clone()
    //     .query(format!("{}", CommitStatement))
    //     .await
    //     .unwrap();
    //
    let c = Company::schema();
    let select_query = select(All)
        .from(&SurrealId::try_from("company:2").unwrap())
        .where_(c.tags.any_like("foo"))
        .parallel();
    // .return_one(db.clone())
    // .await
    // .unwrap();

    dbg!(select_query.clone().get_bindings());
    dbg!(select_query.clone().to_raw());
    assert_eq!(select_query.clone().build().to_string(), "dfd");
    // println!("BindSel {:?}", select_query.get_bindings());
    let one_ = select_query
        .return_one::<Company>(db.clone())
        .await
        .unwrap();
    println!("SSSSSSS {:?}", one_);

    println!(
        "SSSSSSS {:?}",
        select_query
            .return_many::<Company>(db.clone())
            .await
            .unwrap()
    );

    let ref select_query = select(All)
        .from(Company::get_table_name())
        .where_(c.tags.any_like("foo"))
        .timeout(Duration::from_secs(20))
        .parallel();

    println!("BindSel {:?}", select_query.get_bindings());
    println!(
        "SSSSSSS {:?}",
        select_query
            .return_many::<Company>(db.clone())
            .await
            .unwrap()
    );
    // TODO: The fall back to return_one if list returned not working. Investigate.
    // let results = insert::<GenZCompany>(select_query)
    //     .return_one(db.clone())
    //     .await
    //     .unwrap();
    let results: Vec<GenZCompany> = insert(select_query).return_many(db.clone()).await.unwrap();

    let results = insert::<GenZCompany>(select_query)
        .return_many(db.clone())
        .await
        .unwrap();

    let results: Vec<GenZCompany> = insert(select_query).return_many(db.clone()).await.unwrap();
    // let results = insert::<GenZCompany>(select_query)
    //     .return_many(db.clone())
    //     // .return_many_explicit::<Vec<GenZCompany>>(db.clone())
    //     .await
    //     .unwrap();

    insta::assert_debug_snapshot!(results);
    Ok(())
}

