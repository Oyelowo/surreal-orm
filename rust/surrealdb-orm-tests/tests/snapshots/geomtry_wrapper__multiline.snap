---
source: surrealdb-orm-tests/tests/geomtry_wrapper.rs
expression: company
---
{"id":"company:lowo","nam":"285cfebe-a7f2-4100-aeb3-7f73998fff02","name":"Mana Inc.","founded":"1967-05-03T00:00:00Z","founders":[{"name":"John Doe"},{"name":"Jane Doe"}],"tags":["foo","bar"],"home":{"type":"MultiLineString","coordinates":[[[0.0,0.0],[1.0,1.0],[2.0,2.0]],[[3.0,3.0],[4.0,4.0],[5.0,5.0]]]}}
